### PR TITLE
Fix apiservice bats test

### PR DESCRIPTION
**Pull request by tiffanny29631**
This change applies to the recent CI failures in mono repo test under GKE environment.

Also updates the Gcloud CLI used in test containers and installs gke-gcloud-auth-plugin.

### CODE REVIEW

File: e2e/testcases/apiservice.bats [hunk 1]
```diff
@@ -13,17 +13,32 @@ YAML_DIR=${BATS_TEST_DIRNAME}/../testdata/apiservice
 test_setup() {
   setup::git::initialize
   setup::git::commit_minimal_repo_contents
+  wait::for -t 120 -- nomos::repo_synced
```
Comment (nan-yu) [inline e2e/testcases/apiservice.bats:16]: In the end of `setup::git::commit_minimal_repo_contents`, it waits for the repo to be synced, so the additional wait is unnecessary.

File: e2e/testcases/apiservice.bats [hunk 2]
```diff
@@ -13,17 +13,32 @@ YAML_DIR=${BATS_TEST_DIRNAME}/../testdata/apiservice
...
+kubectl delete -f "${YAML_DIR}/apiservice.yaml" --ignore-not-found
+  git::rm acme/namespaces/custom-metrics/namespace.yaml
+  git::rm acme/namespaces/custom-metrics/namespace-custom-metrics.yaml
+  git::commit
+  kubectl delete -f "${YAML_DIR}/namespace.yaml" --ignore-not-found
+  kubectl delete -f "${YAML_DIR}/namespace-custom-metrics.yaml" --ignore-not-found
+  kubectl delete -f "${YAML_DIR}/rbac.yaml" --ignore-not-found
+
   setup::git::remove_all acme
```
Comment (nan-yu) [inline e2e/testcases/apiservice.bats:31]: `setup::git::remove_all acme` deletes all cluster and namespace resources. I don't think you need to delete them explicitly.

File: e2e/testcases/apiservice.bats [hunk 3]
```diff
@@ -13,17 +13,32 @@ YAML_DIR=${BATS_TEST_DIRNAME}/../testdata/apiservice
...
 }
 
 @test "Create APIService and endpoint in same commit" {
+  kubectl apply -f "${YAML_DIR}/namespace.yaml"
+  kubectl apply -f "${YAML_DIR}/rbac.yaml"
```
Comment (nan-yu) [inline e2e/testcases/apiservice.bats:36]: Why do you apply them manually using kubectl? This testcase is supposed to test the resources being managed by Config Sync.

File: e2e/testcases/apiservice.bats [hunk 4]
```diff
@@ -13,17 +13,32 @@ YAML_DIR=${BATS_TEST_DIRNAME}/../testdata/apiservice
...
+
   debug::log "Creating commit with APIService and Deployment"
+  git::add \
+      "${YAML_DIR}/namespace.yaml" \
```
Comment (nan-yu) [inline e2e/testcases/apiservice.bats:40]: This file is already included in `namespace-custom-metrics.yaml`.

File: e2e/testcases/apiservice.bats [hunk 5]
```diff
@@ -48,8 +63,11 @@ test_teardown() {
     acme/namespaces/resilient/namespace-resilient.yaml
   git::commit
 
+  debug::log "Adding APIService backend"
+  kubectl apply -f "${YAML_DIR}/namespace.yaml"
+  kubectl apply -f "${YAML_DIR}/rbac.yaml"
+  kubectl apply -f "${YAML_DIR}/namespace-custom-metrics.yaml"
```
Comment (nan-yu) [inline e2e/testcases/apiservice.bats:69]: What is the intention of applying these resources manually? The purpose of this test case is to verify that config-sync mono repo mode can survive with a bad apiservice.

File: e2e/testcases/apiservice_test.go [hunk 6]
```diff
@@ -37,6 +37,12 @@ func TestCreateAPIServiceAndEndpointInTheSameCommit(t *testing.T) {
 		if t.Failed() {
 			nt.PodLogs(adapterNamespace, adapterName, "pod-custom-metrics-stackdriver-adapter", true)
 		}
+		if !nt.MultiRepo {
+			nt.MustKubectl("delete", "-f", "../testdata/apiservice/apiservice.yaml", "--ignore-not-found")
+			nt.MustKubectl("delete", "-f", "../testdata/apiservice/rbac.yaml", "--ignore-not-found")
+			nt.MustKubectl("delete", "-f", "../testdata/apiservice/namespace-custom-metrics.yaml", "--ignore-not-found")
+			nt.MustKubectl("delete", "-f", "../testdata/apiservice/namespace.yaml", "--ignore-not-found")
+		}
```
Comment (nan-yu) [inline e2e/testcases/apiservice_test.go:45]: Why not remove them from the repository and let Config Sync prune them?

File: e2e/testcases/apiservice_test.go [hunk 7]
```diff
@@ -57,9 +63,6 @@ func TestCreateAPIServiceAndEndpointInTheSameCommit(t *testing.T) {
 	nt.WaitForRepoSyncs()
 
 	nt.RootRepos[configsync.RootSyncName].Remove("acme/namespaces/custom-metrics/namespace-custom-metrics.yaml")
```
Comment (nan-yu) [inline e2e/testcases/apiservice_test.go:65]: If you don't plan to commit the change, you should remove `nt.RootRepos.Remove` as well.

**Review by nan-yu — CHANGES_REQUESTED** [review]

File: e2e/testcases/apiservice.bats (hunk 1 above)
Comment (tiffanny29631) [reply e2e/testcases/apiservice.bats]: Done! Didn't notice that.

File: e2e/testcases/apiservice.bats (hunk 2 above)
Comment (tiffanny29631) [reply e2e/testcases/apiservice.bats:20]: Done!

File: e2e/testcases/apiservice.bats (hunk 3 above)
Comment (tiffanny29631) [reply e2e/testcases/apiservice.bats]: Moved into commit by Config Sync. To not having to breakdown the test data and create more artifacts I'm using unstructured source format for this test bats.

File: e2e/testcases/apiservice.bats (hunk 4 above)
Comment (tiffanny29631) [reply e2e/testcases/apiservice.bats:26]: I split it out to bring more flexibility when applying resources.

File: e2e/testcases/apiservice.bats (hunk 5 above)
Comment (tiffanny29631) [reply e2e/testcases/apiservice.bats:56]: The change in https://github.com/GoogleContainerTools/kpt-config-sync/pull/28 make errors like 503 blocking, where Config Sync will always retry syncing on bad APIs, thus previous test setup will timeout. Manually applying the deployment a while after APIService is deployed is to mock a temporary flaky API communication.

File: e2e/testcases/apiservice_test.go (hunk 6 above)
Comment (tiffanny29631) [reply e2e/testcases/apiservice_test.go]: Don't know why I added them, they are getting removed from git through Config Sync.

File: e2e/testcases/apiservice_test.go (hunk 7 above)
Comment (tiffanny29631) [reply e2e/testcases/apiservice_test.go:59]: It is getting committed in the second line from here.

File: e2e/testcases/apiservice.bats [hunk 8]
```diff
@@ -11,19 +11,23 @@ load "../lib/wait"
 YAML_DIR=${BATS_TEST_DIRNAME}/../testdata/apiservice
 
 test_setup() {
+  kubectl apply -f "${MANIFEST_DIR}/source-format_unstructured.yaml"
```
[the remark this answers is not in the document]
Comment (tiffanny29631) [reply e2e/testcases/apiservice.bats:14]: Added revert to hierarchical mode in test_testdown()

File: e2e/testcases/apiservice.bats [hunk 9]
```diff
@@ -11,19 +11,23 @@ load "../lib/wait"
...
+  nomos::restart_pods
   setup::git::initialize
-  setup::git::commit_minimal_repo_contents
 }
 
 test_teardown() {
-  git::rm acme/cluster/apiservice.yaml
-  git::commit
-  kubectl delete -f "${YAML_DIR}/apiservice.yaml" --ignore-not-found
```
[the remark this answers is not in the document]
Comment (tiffanny29631) [reply e2e/testcases/apiservice.bats:21]: Resumed

File: e2e/testcases/apiservice.bats (hunk 5 above)
Comment (tiffanny29631) [reply e2e/testcases/apiservice.bats:76]: Added comments to explain the manual application and cleanups during test teardown.

File: e2e/testcases/apiservice.bats [hunk 10]
```diff
@@ -41,15 +45,15 @@ test_teardown() {
...
   git::commit
-
+  debug::log "Adding APIService backend"
+  kubectl apply -f "${YAML_DIR}/namespace.yaml"
+  kubectl apply -f "${YAML_DIR}/rbac.yaml"
+  kubectl apply -f "${YAML_DIR}/namespace-custom-metrics.yaml"
   debug::log "Waiting for nomos to stabilize"
   wait::for -t 240 -- nomos::repo_synced
-
-  kubectl delete -f "${YAML_DIR}/apiservice.yaml" --ignore-not-found
```
[the remark this answers is not in the document]
Comment (tiffanny29631) [reply e2e/testcases/apiservice.bats:54]: Resumed

File: e2e/testcases/apiservice_test.go (hunk 7 above)
Comment (tiffanny29631) [reply e2e/testcases/apiservice_test.go:62]: Discussed: let Config Sync remove the test APISerivce first so it's not blocked by Discovery failure later when cleanup rest of repo.

File: e2e/testcases/apiservice.bats [hunk 11]
```diff
@@ -38,18 +56,24 @@ test_teardown() {
   wait::for -t 240 -- nomos::repo_synced
 }
 
-@test "importer and syncer resilient to bad APIService" {
+@test "importer and syncer resilient to flaky APIService" {
```
Comment (sdowell) [inline e2e/testcases/apiservice.bats:59]: Is the bats test still needed given we have an equivalent go test?

Comment (tiffanny29631) [reply e2e/testcases/apiservice.bats:59]: So far it's still needed, the mono repo test on GKE clusters are still running bats tests and this is the quickest way to bring CI back to green. We are going to fully deprecate the bats tests in coming Fixit.

File: e2e/testcases/apiservice.bats (hunk 8 above)
[the remark this answers is not in the document]
Comment (tiffanny29631) [reply e2e/testcases/apiservice.bats:14]: Also doing cleanup to the manually applied resources here, plus letting Config Sync remove the test APIService first so it's not blocked by discovery failures when trying to cleanup acme.

[the remark this answers is not in the document]
Comment (nan-yu) [reply e2e/testcases/apiservice.bats:14]: ACK

File: e2e/testcases/apiservice.bats [hunk 12]
```diff
@@ -38,18 +56,24 @@ test_teardown() {
...
+  # Create an APIService without backend so that Config Sync DiscoveryClient will
+  # get service unavailable errors when trying to parse resource
```
Comment (nan-yu) [inline e2e/testcases/apiservice.bats:61]: Is the service unavailable error visible to the user? Can we add a validation for that if possible?

If we don't add any validation before restoring the backend, it is basically the same as the first test case.

File: e2e/testcases/apiservice_test.go [hunk 13]
```diff
@@ -52,36 +52,40 @@ func TestCreateAPIServiceAndEndpointInTheSameCommit(t *testing.T) {
...
+		nt.MustKubectl("delete", "-f", "../testdata/apiservice/namespace-custom-metrics.yaml", "--ignore-not-found")
 		if t.Failed() {
 			nt.PodLogs(adapterNamespace, adapterName, "pod-custom-metrics-stackdriver-adapter", true)
 		}
 	})
 
+	nt.T.Log("Adding APIService with backend temporarily unavailable to trigger discovery errors")
+	nt.MustKubectl("apply", "-f", "../testdata/apiservice/apiservice.yaml")
+
 	nt.T.Log("Creating commit with resources")
```
Comment (nan-yu) [inline e2e/testcases/apiservice_test.go:82]: Similarly as the bats test, can we validate the backend unavailable error?

File: e2e/testcases/apiservice.bats [hunk 14]
```diff
@@ -38,18 +56,27 @@ test_teardown() {
...
   kubectl apply -f "${YAML_DIR}/apiservice.yaml"
 
-  debug::log "Creating commit with Deployment"
+  debug::log "Creating a commit with some test resources"
   git::add \
     "${YAML_DIR}/namespace-resilient.yaml" \
     acme/namespaces/resilient/namespace-resilient.yaml
   git::commit
 
+  debug::log "Waiting for APIServerError to be present in Rootsync status"
```
Comment (nan-yu) [inline e2e/testcases/apiservice.bats:71]: nit: The `nomos::source_error_code` function works for both mono-repo and multi-repo. Suggest removing `Rootsync` in the comment as that is specific to the multi-repo mode.

File: e2e/testcases/apiservice_test.go [hunk 15]
```diff
@@ -52,36 +53,48 @@ func TestCreateAPIServiceAndEndpointInTheSameCommit(t *testing.T) {
...
+	nt.T.Log("Creating commit with test resources")
+	nt.RootRepos[configsync.RootSyncName].Copy("../testdata/apiservice/namespace-resilient.yaml", "acme/namespaces/resilient/namespace.yaml")
+	//nt.RootRepos[configsync.RootSyncName].Copy("../testdata/apiservice/apiservice.yaml", "acme/cluster/apiservice.yaml")
 	nt.RootRepos[configsync.RootSyncName].CommitAndPush("add testing resources")
 
-	nt.T.Log("Adding backend")
+	nt.T.Log("Waiting for APIServerError to be present in Rootsync status")
+	if !nt.MultiRepo {
+		nt.WaitForRepoSourceError(status.APIServerErrorCode)
+	}
```
Comment (nan-yu) [inline e2e/testcases/apiservice_test.go:93]: else?

The test runs on both mono-repo mode and multi-repo mode.

File: e2e/testcases/apiservice_test.go [hunk 16]
```diff
@@ -52,36 +53,48 @@ func TestCreateAPIServiceAndEndpointInTheSameCommit(t *testing.T) {
...
 	})
 
-	nt.T.Log("Creating commit with resources")
-	nt.RootRepos[configsync.RootSyncName].Copy("../testdata/apiservice/rbac.yaml", "acme/cluster/rbac.yaml")
+	nt.T.Log("Adding APIService with backend temporarily unavailable to trigger discovery errors")
+	nt.MustKubectl("apply", "-f", "../testdata/apiservice/apiservice.yaml")
+
+	nt.T.Log("Creating commit with test resources")
+	nt.RootRepos[configsync.RootSyncName].Copy("../testdata/apiservice/namespace-resilient.yaml", "acme/namespaces/resilient/namespace.yaml")
+	//nt.RootRepos[configsync.RootSyncName].Copy("../testdata/apiservice/apiservice.yaml", "acme/cluster/apiservice.yaml")
```
[the remark this answers is not in the document]
Comment (tiffanny29631) [reply e2e/testcases/apiservice_test.go]: Done

File: e2e/testcases/apiservice.bats (hunk 14 above)
Comment (tiffanny29631) [reply e2e/testcases/apiservice.bats]: Done

File: e2e/testcases/apiservice_test.go (hunk 15 above)
Comment (tiffanny29631) [reply e2e/testcases/apiservice_test.go:107]: Just confirmed in debugging env that under multirepo mode the test resource can be correctly reconciled even with bad APIService. Added RG status check for the test resource to make sure.

File: e2e/testcases/apiservice_test.go [hunk 17]
```diff
@@ -52,36 +53,62 @@ func TestCreateAPIServiceAndEndpointInTheSameCommit(t *testing.T) {
...
+			err = nt.Validate("resilient", "", &corev1.Namespace{},
+				nomostest.StatusEquals(nt, kstatus.CurrentStatus))
+			if err != nil {
+				return err
+			}
+			return nil
+		})
+		if e != nil {
+			nt.T.Fatal("validate failed test resource to have status CURRENT")
+		}
```
Comment (nan-yu) [inline e2e/testcases/apiservice_test.go:103]: Just to confirm, if the APIService is unavailable in the multi-repo mode, there won't be any effect to the sync and reconciliation status. Basically, user won't notice any difference. Is that the expected behavior?

Comment (tiffanny29631) [reply e2e/testcases/apiservice_test.go:102]: In this test scenario where APIService is unavailable what you described is correct. I was also working on producing a case where multi-repo mode fails to reconcile due to timeout discovery failure until this bats test starts failing. Will switch back to that part once CI is fixed.

Comment (nan-yu) [reply e2e/testcases/apiservice_test.go:102]: SG

**Review by nan-yu — APPROVED** [review]
/lgtm

**Review by nan-yu — APPROVED** [review]
/lgtm

**Pull request merged**